crates/node/src/shell/finalize_block.rs

```diff
@@ -340,29 +338,50 @@ where
         &mut self,
         response: &mut shim::response::FinalizeBlock,
         extended_dispatch_result: std::result::Result<
-            namada_sdk::tx::data::ExtendedTxResult<protocol::Error>,
+            namada_sdk::tx::data::TxResult<protocol::Error>,
             DispatchError,
         >,
         tx_data: TxData<'_>,
         mut tx_logs: TxLogs<'_>,
     ) -> Option<WrapperCache> {
         match extended_dispatch_result {
-            Ok(extended_tx_result) => match tx_data.tx.header.tx_type {
+            Ok(tx_result) => match tx_data.tx.header.tx_type {
                 TxType::Wrapper(_) => {
+                    // Commit any changes brought in by the optional fee-paying
+                    // tx to ensure fee payment
                     self.state.write_log_mut().commit_batch_and_current_tx();
 
-                    // Return withouth emitting any events
+                    // Emit the events of the fee-paying inner tx. This tx has
+                    // just been committed so we need to ensure the relative
+                    // events are emitted
+                    let events = tx_result.0.first_key_value().map_or_else(
```

Comment:
Again, I find this assumption dangerous. Maybe we can soup up the type to keep thses events under a separate field

Reply:
We can but we need to handle the case of a failing atomic batch for which the fee paying tx should be committed anyway (and the relative events should be emitted). My reason to emit the events here is that in the line above we commit the storage changes and I wanted to emit the events at the same time.

Also, by emitting the events here, the masp events are already sorted in the block results (without the need to look at the masp fee payment attribute), so that clients don't need to reorder them.